include/bamit/sample_functions.hpp

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <algorithm> // For std::sort
+#include <random> // For std::random_device
+#include <numeric> // For std::reduce
+#include <cmath> // For std::sqrt and std::pow
+
+#include <bamit/Record.hpp>
+#include <bamit/IntervalNode.hpp>
+
+#include <seqan3/io/sam_file/input.hpp>
+#include <seqan3/core/debug_stream.hpp>
+
+namespace bamit
+{
+
+/*! Contains statistics which can be calculated over a sample. */
+struct EstimationResult
+{
+    double mean{0}, median{0}, mode{0}, sd{0}, variance{0};
+
+    /*! Print out the statistics stored within. */
+    void print()
+    {
+        seqan3::debug_stream << "Mean: " << mean
+                             << "\nMedian: " << median
+                             << "\nMode: " << mode
+                             << "\nSD: " << sd
+                             << "\nVariance: " << variance << "\n";
+    }
+};
+
+/*!
+   \brief A function to sample read depth from an alignment file, given the input file, index, and number of positions.
+   \param input_file The input alignment file in sam/bam format.
+   \param bamit_index The vector of indices for each chromosome.
+   \param sample_value The number of positions to sample.
+
+   \details This function samples an alignment file at `sample_value` different random positions according to a uniform
+            distribution and gives the read depth of the file as a series of statistics with mean, median, mode, sd and
+            variance stored in an EstimationResult struct.
+
+   \return Returns a struct containing statistics over the sampled points.
+ */
+template <typename traits_type, typename fields_type, typename format_type>
+inline EstimationResult sample_read_depth(seqan3::sam_file_input<traits_type, fields_type, format_type> & input_file,
+                                          std::vector<std::unique_ptr<IntervalNode>> const & bamit_index,
+                                          uint64_t const & sample_value)
+    {
+        if (sample_value <= 1) throw std::invalid_argument("sample_value must be greater than 1.");
+        auto & header = input_file.header(); // Obtain the header from the input file.
+        // Intialize vector of read depths at # of positions defined by sample_value.
+        std::vector<uint64_t> read_depths(sample_value);
+        std::map<double, uint64_t> depth_counts{}; // Counts how often each depth occurs.
+        std::random_device rd; // obtain a random number from hardware
+        std::mt19937 gen(rd()); // seed the generator
```

Comment:
Sorry I have another comment: Is it necessary that we take a true random number, or can we fix the seed and achieve a deterministic result? I just ran across this problem in a completely different place and would like it if we could specify a fixed seed here as well.

Reply:
Mmmm I mean I guess that is possible, but I would think more randomness is better. Otherwise you could have some unlucky sampling or something, no?

Reply:
Yes, but the user does not know whether a bad seed has just been selected. He can then only call again and hope for a better result. Is it better to give the user the possibility to set the seed (optional)?

Just as a background I have stumbled in iGenVar here: [line 150](https://github.com/seqan/iGenVar/blob/master/src/modules/clustering/hierarchical_clustering_method.cpp) about it and will now change this as in SVIM [line 129](https://github.com/eldariont/svim/blob/master/src/svim/SVIM_clustering.py) to a fixed seed.

Reply:
One possibility I see, you could make the seed specifiable, and if it is not specified, then random_device is used. Then you have a deterministic result if you want. But I think you always want to have the latter.